### PR TITLE
Update aws-lambda-java-core version and fix bug in the multiple skills usecase

### PR DESCRIPTION
**Pull request by prakshai**
1. Update aws-lambda-java-core version with related code changes
2. Fix bug to ensure multiple skills usecase behaves as expected
3. If skillId in the ASK request doesn't match the one used when initializing the skill, let the fallthrough happen to the next skill.

## Description
1. The change updates aws-lambda-java-core and applies code changes required to build successfully post the version upgrade.
2. We have a usecase where we chose to host multiple skills as part of the same application using this skills stream handler constructor: https://github.com/alexa/alexa-skills-kit-sdk-for-java/blob/2.0.x/ask-sdk-lambda-support/src/com/amazon/ask/SkillStreamHandler.java#L65. When we try to invoke the second skill within that handler we see an exception "com.amazon.ask.exception.UnhandledSkillException: Unhandled exception" in our logs. Upon debugging, the issue seems to be when running multiple skills, the second skill's (possibly any skills apart from the first skill in the list) handlers are ignored.
3. If skillId in the ASK request doesn't match the one used when initializing the skill, let the fallthrough happen to the next skill. Only fail will all skillIds in the list don't match the skillId in the ASK request. Fixing the workflow [here](https://github.com/alexa/alexa-skills-kit-sdk-for-java/blob/d60b98038a2e20ae58e8652521e6cf6f7d98b034/ask-sdk-core/src/com/amazon/ask/CustomSkill.java#L121-L123) by not throwing and exception but returning a null to be handled as appropriate.

## Motivation and Context
When trying to build with this project as a dependency on Android platform the build fails with "The prefix "xsi" for attribute "xsi:schemaLocation" associated with an element type "project" is not bound". 
Root cause: without these declarations, the XML file cannot be validated correctly. 
Refer: https://github.com/aws/aws-lambda-java-libs/issues/89

As a work around, for now, currently we are excluding the aws-lambda-java-core dependency and forcing the new version to be used during build. But this proposed change will allow us to clean up that work around.

## Testing
Built using mvn clean install. Built successfully.
Tested by running two skills within the same application and both skills executed as expected.

## Types of changes

- [x ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x ] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have read the **README** document
- [X ] I have added tests to cover my changes
- [x ] All new and existing tests passed

## License
- [ x] By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

[issues]: https://github.com/alexa/alexa-skills-kit-sdk-for-java/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

### CODE REVIEW

**Comment by prakshai** [comment]
Some of the files that are failing checkstyle aren't even the ones I edited. Regardless, will be more than happy to fix them if there are some instructions on how to fix them.

File: ask-sdk-core/src/com/amazon/ask/CustomSkill.java [hunk 1]
```diff
@@ -119,7 +127,8 @@ protected ResponseEnvelope invoke(final UnmarshalledRequest<RequestEnvelope> unm
         JsonNode requestEnvelopeJson = unmarshalledRequest.getRequestJson();
 
         if (skillId != null && !requestEnvelope.getContext().getSystem().getApplication().getApplicationId().equals(skillId)) {
-            throw new AskSdkException("AlexaSkill ID verification failed.");
+            LOGGER.debug("AlexaSkill ID verification failed.");
```
Comment (sattpat) [inline ask-sdk-core/src/com/amazon/ask/CustomSkill.java]: nit.. add the expected and actual skill id to the log message here

**Comment by sattpat** [comment]
The PR looks to be missing tests for a scenario where all the skills in the collection are visited but none can handle the skill request and ends up throwing an exception?

**Review by sattpat — CHANGES_REQUESTED** [review]
The PR looks to be missing tests for a scenario where all the skills in the collection are visited but none can handle the skill request and ends up throwing an exception?

**Comment by prakshai** [comment]
> The PR looks to be missing tests for a scenario where all the skills in the collection are visited but none can handle the skill request and ends up throwing an exception?

I thought this should cover for that: https://github.com/alexa/alexa-skills-kit-sdk-for-java/blob/2.0.x/ask-sdk-lambda-support/tst/com/amazon/ask/SkillStreamHandlerTest.java#L91-L95

File: ask-sdk-core/src/com/amazon/ask/CustomSkill.java (hunk 1 above)
Comment (prakshai) [reply ask-sdk-core/src/com/amazon/ask/CustomSkill.java]: Will update with a new commit

File: ask-sdk-lambda-support/src/com/amazon/ask/SkillStreamHandler.java [hunk 2]
```diff
@@ -80,8 +81,8 @@ public SkillStreamHandler(final AlexaSkill... skills) {
     public final void handleRequest(final InputStream input, final OutputStream output, final Context context)
             throws IOException {
         byte[] inputBytes = IOUtils.toByteArray(input);
-        for (AlexaSkill skill : skills) {
-            SkillResponse response = skill.execute(new BaseSkillRequest(inputBytes), context);
+        for (int i = 0; i < skills.size(); ++i) {
```
Comment (sattpat) [inline ask-sdk-lambda-support/src/com/amazon/ask/SkillStreamHandler.java]: Nit: why not iterate over using foreach?

Comment (prakshai) [reply ask-sdk-lambda-support/src/com/amazon/ask/SkillStreamHandler.java]: Ha, my bad, changes while testing

**Review by sattpat — APPROVED** [review]

File: ask-sdk-lambda-support/src/com/amazon/ask/SkillStreamHandler.java [hunk 3]
```diff
@@ -27,6 +27,7 @@
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
```
Comment (nikhilym) [inline ask-sdk-lambda-support/src/com/amazon/ask/SkillStreamHandler.java]: Guess you forgot removing the `import` here

File: ask-sdk-runtime/src/com/amazon/ask/request/dispatcher/impl/BaseRequestDispatcher.java [hunk 4]
```diff
@@ -145,8 +145,8 @@ private Output doDispatch(final Input input) throws Exception {
             LOGGER.debug("Found matching handler");
         } else {
             String message = "Unable to find a suitable request handler";
-            LOGGER.error(message);
-            throw new AskSdkException(message);
+            LOGGER.debug(message);
+            return null;
```
Comment (nikhilym) [inline ask-sdk-runtime/src/com/amazon/ask/request/dispatcher/impl/BaseRequestDispatcher.java:149]: By changing this behavior, how does it affect the case where the request handler is not present? Does the service return a response for that or just fail? How will the skill developer know the issue?

Comment (sattpat) [reply ask-sdk-runtime/src/com/amazon/ask/request/dispatcher/impl/BaseRequestDispatcher.java:149]: So we will now rely on this exception to be thrown - https://github.com/alexa/alexa-skills-kit-sdk-for-java/blob/3424376c148b88a5a7b82311c995151154f879b9/ask-sdk-lambda-support/src/com/amazon/ask/SkillStreamHandler.java#L92

This will be thrown once we have iterated over the collection of skills.

File: ask-sdk-lambda-support/src/com/amazon/ask/SkillStreamHandler.java (hunk 3 above)
Comment (prakshai) [reply ask-sdk-lambda-support/src/com/amazon/ask/SkillStreamHandler.java]: Yes. Fixing it.

File: ask-sdk-runtime/src/com/amazon/ask/request/dispatcher/impl/BaseRequestDispatcher.java (hunk 4 above)
Comment (prakshai) [reply ask-sdk-runtime/src/com/amazon/ask/request/dispatcher/impl/BaseRequestDispatcher.java:149]: This is I think what it should have been if this worked as expected.

File: ask-sdk-lambda-support/src/com/amazon/ask/SkillStreamHandler.java (hunk 3 above)
Comment (prakshai) [reply ask-sdk-lambda-support/src/com/amazon/ask/SkillStreamHandler.java]: Fixed

File: ask-sdk-runtime/src/com/amazon/ask/request/dispatcher/impl/BaseRequestDispatcher.java (hunk 4 above)
Comment (nikhilym) [reply ask-sdk-runtime/src/com/amazon/ask/request/dispatcher/impl/BaseRequestDispatcher.java:149]: > So we will now rely on this exception to be thrown -
> 
> https://github.com/alexa/alexa-skills-kit-sdk-for-java/blob/3424376c148b88a5a7b82311c995151154f879b9/ask-sdk-lambda-support/src/com/amazon/ask/SkillStreamHandler.java#L92
> 
> This will be thrown once we have iterated over the collection of skills.

That seems to be for `lambda`. How about for the skill servlet package? Is it handled similarly in that workflow as well?

Comment (sattpat) [reply ask-sdk-runtime/src/com/amazon/ask/request/dispatcher/impl/BaseRequestDispatcher.java:149]: Thats a great point! I see that the servlet is not taking a skill collection https://github.com/alexa/alexa-skills-kit-sdk-for-java/blob/3424376c148b88a5a7b82311c995151154f879b9/ask-sdk-servlet-support/src/com/amazon/ask/servlet/SkillServlet.java#L79

File: ask-sdk-servlet-support/tst/com/amazon/ask/servlet/SkillServletTest.java [hunk 5]
```diff
@@ -65,18 +65,6 @@ public void setup() {
         skill = mock(Skill.class);
     }
 
-    @Test
-    public void doPost_passedVerification_responseSuccess() throws Exception {
```
Comment (sattpat) [inline ask-sdk-servlet-support/tst/com/amazon/ask/servlet/SkillServletTest.java:69]: Can you add a test here, where the skill.execute returns null and the asksdkexception gets thrown by the skillservlet class? The current tests mocks the skill.execute to throw an exception which is not true anymore since we have changed the logic in the base dispatcher

Comment (prakshai) [reply ask-sdk-servlet-support/tst/com/amazon/ask/servlet/SkillServletTest.java:69]: Added test for this specific case.

**Review by sattpat — APPROVED** [review]

**Review by sattpat — CHANGES_REQUESTED** [review]
I see that the latest travis build is failing because of an unused import

**Comment by prakshai** [comment]
> I see that the latest travis build is failing because of an unused import

Pushed a change to fix it

File: ask-sdk-local-debug/src/com/amazon/ask/localdebug/lambdaContext/DebugLambdaContext.java [hunk 6]
```diff
@@ -60,6 +60,16 @@ public String getFunctionName() {
     /**
      * Defaults to returning null cognito identity.
      */
+    @Override
```
Comment (nikhilym) [inline ask-sdk-local-debug/src/com/amazon/ask/localdebug/lambdaContext/DebugLambdaContext.java:63]: Can you please move the doc string to `getIdentity` function below?

File: ask-sdk-local-debug/src/com/amazon/ask/localdebug/lambdaContext/DebugLambdaContext.java [hunk 7]
```diff
@@ -60,6 +60,16 @@ public String getFunctionName() {
...
+    public String getFunctionVersion() {
+        return null;
+    }
+
+    @Override
```
Comment (nikhilym) [inline ask-sdk-local-debug/src/com/amazon/ask/localdebug/lambdaContext/DebugLambdaContext.java:71]: Add a doc string here for fixing travis issues

File: ask-sdk-local-debug/src/com/amazon/ask/localdebug/lambdaContext/DebugLambdaContext.java (hunk 6 above)
Comment (nikhilym) [reply ask-sdk-local-debug/src/com/amazon/ask/localdebug/lambdaContext/DebugLambdaContext.java:63]: Also add doc string for this new function.

**Review by sattpat — CHANGES_REQUESTED** [review]
You need to add javadoc getIdentity and getInvokedFunctionArn in Class 'DebugLambdaContext'

**Review by nikhilym — CHANGES_REQUESTED** [review]

File: ask-sdk-local-debug/src/com/amazon/ask/localdebug/lambdaContext/DebugLambdaContext.java (hunk 7 above)
Comment (prakshai) [reply ask-sdk-local-debug/src/com/amazon/ask/localdebug/lambdaContext/DebugLambdaContext.java:71]: updated

File: ask-sdk-local-debug/src/com/amazon/ask/localdebug/lambdaContext/DebugLambdaContext.java (hunk 6 above)
Comment (prakshai) [reply ask-sdk-local-debug/src/com/amazon/ask/localdebug/lambdaContext/DebugLambdaContext.java:63]: Updated

**Review by sattpat — APPROVED** [review]
:shipit:

**Review by nikhilym — APPROVED** [review]
LGTM :shipit:
